File: lib/observability.js
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Neil Blackman
//
// Privacy-safe, low-cardinality MCP metrics and completion logs. This module never accepts
// request arguments, identities, authorization headers, artifact contents, or raw resource URIs.
import { randomUUID } from "node:crypto";

const DURATION_BUCKETS_SECONDS = [0.01, 0.05, 0.1, 0.25, 0.5, 1, 5];
const KNOWN_METHODS = new Set([
  "initialize",
  "server/discover",
  "tools/list",
  "tools/call",
  "resources/list",
  "resources/templates/list",
  "resources/read",
  "tasks/get",
  "tasks/update",
  "tasks/cancel"
]);
const KNOWN_TOOLS = new Set([
  "publish_artifact",
  "publish_bundle",
  "list_artifacts",
  "delete_artifact",
  "update_artifact",
  "set_visibility",
  "list_categories",
  "set_category",
  "create_category",
  "delete_category",
  "list_revisions",
  "create_share",
  "list_shares",
  "revoke_share",
  "artifact_stats",
  "submit_feedback",
  "list_feedback",
  "resolve_feedback",
  "reopen_feedback",
  "read_artifact",
  "patch_artifact",
  "regenerate_artifact_preview"
]);

function operationFor(method) {
  if (method === "server/discover") return "discovery";
  if (method === "initialize") return "initialization";
  if (method === "tools/list") return "listing";
  if (method === "tools/call") return "tool_call";
  if (method === "resources/list" || method === "resources/templates/list" || method === "resources/read") {
    return "resource";
  }
  if (method === "tasks/get" || method === "tasks/update" || method === "tasks/cancel") return "task";
  return "unknown";
}

function safeName(method, name) {
  if (method === "resources/read") {
    if (name === "ui://artifact-mcp/review") return "review_app";
    if (typeof name === "string" && name.startsWith("artifact://") && name.endsWith("/thumbnail")) {
      return "artifact_thumbnail";
    }
    if (typeof name === "string" && name.startsWith("artifact://")) return "artifact_content";
    return "unknown";
  }
  if (method.startsWith("tasks/")) return "preview_regeneration";
  if (method !== "tools/call") return "none";
  return KNOWN_TOOLS.has(name) ? name : "unknown";
}

export function mcpMetricLabels(protocol, method, name) {
  const safeMethod = KNOWN_METHODS.has(method) ? method : "unknown";
  return {
    protocol: protocol === "2026-07-28" || protocol === "2025-06-18" || protocol === "unsupported"
      ? protocol
      : "unknown",
    operation: operationFor(safeMethod),
    method: safeMethod,
    name: safeName(safeMethod, name)
  };
}

export function mcpProtocolDimension(headers = {}) {
  const value = headers["mcp-protocol-version"];
  if (value === "2026-07-28") return value;
  if (value == null || value === "2025-06-18") return "2025-06-18";
  return "unsupported";
}

export function mcpRequestName(body) {
  const method = body?.method;
  if (method === "resources/read") return body?.params?.uri;
  if (method?.startsWith("tasks/")) return body?.params?.taskId;
  return body?.params?.name;
}

export function mcpResponseOutcome(output) {
  const code = output?.error?.code;
  const message = String(output?.error?.message || "");
  if (code === -32603 && (
    message.includes("output failed validation")
    || message.includes("output schema")
    || message.includes("structured content")
  )) return "output_validation_failure";
  if (code === -32603) return "server_failure";
  if (code === -32602 || code === -32600 || code === -32700) return "validation_failure";
  if (output?.error) return "protocol_error";
  return "success";
}

function resultSizeBucket(bytes) {
  if (bytes <= 1_024) return "le_1k";
  if (bytes <= 16_384) return "le_16k";
  if (bytes <= 262_144) return "le_256k";
  if (bytes <= 1_048_576) return "le_1m";
  return "gt_1m";
}

function metricKey(labels, outcome) {
  return JSON.stringify([
    labels.protocol,
    labels.operation,
    labels.method,
    labels.name,
    outcome
  ]);
}

function emptySeries() {
  return {
    calls: 0,
    durationSumSeconds: 0,
    durationBuckets: Array(DURATION_BUCKETS_SECONDS.length + 1).fill(0),
    resultBytes: 0,
    resultSizeBuckets: new Map([
      ["le_1k", 0],
      ["le_16k", 0],
      ["le_256k", 0],
      ["le_1m", 0],
      ["gt_1m", 0]
    ])
  };
}

export function createMcpTelemetry({
  logger = console,
  now = () => performance.now(),
  createRequestId = () => `mcp_${randomUUID().replaceAll("-", "").slice(0, 16)}`
} = {}) {
  const series = new Map();

  function record(requestId, labels, outcome, durationSeconds, resultBytes) {
    const key = metricKey(labels, outcome);
    const item = series.get(key) || emptySeries();
    item.calls += 1;
    item.durationSumSeconds += durationSeconds;
    const durationBucket = DURATION_BUCKETS_SECONDS.findIndex((bound) => durationSeconds <= bound);
    item.durationBuckets[durationBucket === -1 ? DURATION_BUCKETS_SECONDS.length : durationBucket] += 1;
    item.resultBytes += resultBytes;
    const sizeBucket = resultSizeBucket(resultBytes);
    item.resultSizeBuckets.set(sizeBucket, item.resultSizeBuckets.get(sizeBucket) + 1);
    series.set(key, item);

    logger.info?.("[artifact-mcp] MCP request completed", {
      request_id: requestId,
      ...labels,
      outcome,
      duration_ms: durationSeconds * 1_000,
      result_size_bucket: sizeBucket
    });
  }

  function begin() {
    const requestId = createRequestId();
    const started = now();
    let labels = mcpMetricLabels("unknown");
    let completed = false;
    return {
      requestId,
      setLabels(next) {
        labels = next;
      },
      finish(outcome, resultBytes = 0) {
        if (completed) return;
        completed = true;
        record(requestId, labels, outcome, Math.max(0, now() - started) / 1_000, resultBytes);
      },
      cancel() {
        this.finish("cancelled", 0);
      },
      get completed() {
        return completed;
      }
    };
  }

  function renderPrometheus() {
    let output =
      "# HELP artifact_mcp_requests_total MCP requests by safe protocol dimensions and outcome.\n"
      + "# TYPE artifact_mcp_requests_total counter\n"
      + "# HELP artifact_mcp_request_duration_seconds MCP request duration in seconds.\n"
      + "# TYPE artifact_mcp_request_duration_seconds histogram\n"
      + "# HELP artifact_mcp_result_bytes_total Total serialized MCP response bytes.\n"
      + "# TYPE artifact_mcp_result_bytes_total counter\n"
      + "# HELP artifact_mcp_result_size_bucket_total MCP responses in bounded size bands.\n"
      + "# TYPE artifact_mcp_result_size_bucket_total counter\n";
    for (const [rawKey, item] of [...series.entries()].sort(([left], [right]) => left.localeCompare(right))) {
      const [protocol, operation, method, name, outcome] = JSON.parse(rawKey);
      const labels = `protocol="${protocol}",operation="${operation}",method="${method}",name="${name}",outcome="${outcome}"`;
      output += `artifact_mcp_requests_total{${labels}} ${item.calls}\n`;
      let cumulative = 0;
      DURATION_BUCKETS_SECONDS.forEach((bound, index) => {
        cumulative += item.durationBuckets[index];
        output += `artifact_mcp_request_duration_seconds_bucket{${labels},le="${bound}"} ${cumulative}\n`;
      });
      cumulative += item.durationBuckets.at(-1);
      output += `artifact_mcp_request_duration_seconds_bucket{${labels},le="+Inf"} ${cumulative}\n`;
      output += `artifact_mcp_request_duration_seconds_sum{${labels}} ${item.durationSumSeconds}\n`;
      output += `artifact_mcp_request_duration_seconds_count{${labels}} ${item.calls}\n`;
      output += `artifact_mcp_result_bytes_total{${labels}} ${item.resultBytes}\n`;
      for (const [size, count] of item.resultSizeBuckets) {
        output += `artifact_mcp_result_size_bucket_total{${labels},size="${size}"} ${count}\n`;
      }
    }
    return output;
  }

  return { begin, renderPrometheus };
}

File: src/observability.rs
//! Privacy-safe, low-cardinality MCP telemetry.
//!
//! The recorder intentionally accepts only protocol dimensions, never request arguments,
//! identities, resource URIs, authorization headers, or artifact content. Metrics remain useful
//! to Prometheus while an opaque server-generated request id connects them to the matching
//! structured completion event.

use std::{
    collections::BTreeMap,
    sync::{Arc, Mutex},
    time::Instant,
};

use nanoid::nanoid;

const DURATION_BUCKETS_SECONDS: [f64; 7] = [0.01, 0.05, 0.1, 0.25, 0.5, 1.0, 5.0];

#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
struct MetricKey {
    protocol: &'static str,
    operation: &'static str,
    method: &'static str,
    name: &'static str,
    outcome: &'static str,
}

#[derive(Clone, Copy, Debug)]
pub struct McpMetricLabels {
    pub protocol: &'static str,
    pub operation: &'static str,
    pub method: &'static str,
    pub name: &'static str,
}

impl Default for McpMetricLabels {
    fn default() -> Self {
        Self {
            protocol: "unknown",
            operation: "unknown",
            method: "unknown",
            name: "none",
        }
    }
}

#[derive(Clone, Copy, Debug)]
pub enum McpOutcome {
    Success,
    AuthenticationFailure,
    AuthorizationFailure,
    ValidationFailure,
    OutputValidationFailure,
    ProtocolError,
    ServerFailure,
    Cancelled,
}

impl McpOutcome {
    fn as_str(self) -> &'static str {
        match self {
            Self::Success => "success",
            Self::AuthenticationFailure => "authentication_failure",
            Self::AuthorizationFailure => "authorization_failure",
            Self::ValidationFailure => "validation_failure",
            Self::OutputValidationFailure => "output_validation_failure",
            Self::ProtocolError => "protocol_error",
            Self::ServerFailure => "server_failure",
            Self::Cancelled => "cancelled",
        }
    }
}

#[derive(Default)]
struct Series {
    calls: u64,
    duration_sum_seconds: f64,
    duration_buckets: [u64; DURATION_BUCKETS_SECONDS.len() + 1],
    result_bytes: u64,
    result_size_buckets: [u64; 5],
}

#[derive(Clone, Default)]
pub struct McpTelemetry {
    series: Arc<Mutex<BTreeMap<MetricKey, Series>>>,
}

impl McpTelemetry {
    pub fn begin(&self) -> McpObservation {
        McpObservation {
            telemetry: self.clone(),
            request_id: format!("mcp_{}", nanoid!(16)),
            started: Instant::now(),
            labels: McpMetricLabels::default(),
            completed: false,
        }
    }

    fn record(
        &self,
        request_id: &str,
        labels: McpMetricLabels,
        outcome: McpOutcome,
        duration_seconds: f64,
        result_bytes: usize,
    ) {
        let key = MetricKey {
            protocol: labels.protocol,
            operation: labels.operation,
            method: labels.method,
            name: labels.name,
            outcome: outcome.as_str(),
        };
        let mut series = self
            .series
            .lock()
            .unwrap_or_else(|error| error.into_inner());
        let item = series.entry(key).or_default();
        item.calls = item.calls.saturating_add(1);
        item.duration_sum_seconds += duration_seconds;
        let duration_bucket = DURATION_BUCKETS_SECONDS
            .iter()
            .position(|bound| duration_seconds <= *bound)
            .unwrap_or(DURATION_BUCKETS_SECONDS.len());
        item.duration_buckets[duration_bucket] =
            item.duration_buckets[duration_bucket].saturating_add(1);
        item.result_bytes = item
            .result_bytes
            .saturating_add(u64::try_from(result_bytes).unwrap_or(u64::MAX));
        let result_bucket = match result_bytes {
            0..=1_024 => 0,
            1_025..=16_384 => 1,
            16_385..=262_144 => 2,
            262_145..=1_048_576 => 3,
            _ => 4,
        };
        item.result_size_buckets[result_bucket] =
            item.result_size_buckets[result_bucket].saturating_add(1);
        drop(series);

        tracing::info!(
            target: "artifact_mcp::mcp",
            request_id,
            protocol = labels.protocol,
            operation = labels.operation,
            method = labels.method,
            name = labels.name,
            outcome = outcome.as_str(),
            duration_ms = duration_seconds * 1_000.0,
            result_size_bucket = result_size_bucket(result_bytes),
            "MCP request completed"
        );
    }

    pub fn render_prometheus(&self) -> String {
        let series = self
            .series
            .lock()
            .unwrap_or_else(|error| error.into_inner());
        let mut output = String::from(
            "# HELP artifact_mcp_requests_total MCP requests by safe protocol dimensions and outcome.\n\
             # TYPE artifact_mcp_requests_total counter\n\
             # HELP artifact_mcp_request_duration_seconds MCP request duration in seconds.\n\
             # TYPE artifact_mcp_request_duration_seconds histogram\n\
             # HELP artifact_mcp_result_bytes_total Total serialized MCP response bytes.\n\
             # TYPE artifact_mcp_result_bytes_total counter\n\
             # HELP artifact_mcp_result_size_bucket_total MCP responses in bounded size bands.\n\
             # TYPE artifact_mcp_result_size_bucket_total counter\n",
        );
        for (key, value) in series.iter() {
            let labels = format!(
                "protocol=\"{}\",operation=\"{}\",method=\"{}\",name=\"{}\",outcome=\"{}\"",
                key.protocol, key.operation, key.method, key.name, key.outcome
            );
            output.push_str(&format!(
                "artifact_mcp_requests_total{{{labels}}} {}\n",
                value.calls
            ));
            let mut cumulative = 0_u64;
            for (index, bound) in DURATION_BUCKETS_SECONDS.iter().enumerate() {
                cumulative = cumulative.saturating_add(value.duration_buckets[index]);
                output.push_str(&format!(
                    "artifact_mcp_request_duration_seconds_bucket{{{labels},le=\"{bound}\"}} {cumulative}\n"
                ));
            }
            cumulative =
                cumulative.saturating_add(value.duration_buckets[DURATION_BUCKETS_SECONDS.len()]);
            output.push_str(&format!(
                "artifact_mcp_request_duration_seconds_bucket{{{labels},le=\"+Inf\"}} {cumulative}\n\
                 artifact_mcp_request_duration_seconds_sum{{{labels}}} {}\n\
                 artifact_mcp_request_duration_seconds_count{{{labels}}} {}\n\
                 artifact_mcp_result_bytes_total{{{labels}}} {}\n",
                value.duration_sum_seconds, value.calls, value.result_bytes
            ));
            for (bucket, count) in [
                ("le_1k", value.result_size_buckets[0]),
                ("le_16k", value.result_size_buckets[1]),
                ("le_256k", value.result_size_buckets[2]),
                ("le_1m", value.result_size_buckets[3]),
                ("gt_1m", value.result_size_buckets[4]),
            ] {
                output.push_str(&format!(
                    "artifact_mcp_result_size_bucket_total{{{labels},size=\"{bucket}\"}} {count}\n"
                ));
            }
        }
        output
    }
}

pub struct McpObservation {
    telemetry: McpTelemetry,
    request_id: String,
    started: Instant,
    labels: McpMetricLabels,
    completed: bool,
}

impl McpObservation {
    pub fn request_id(&self) -> &str {
        &self.request_id
    }

    pub fn set_labels(&mut self, labels: McpMetricLabels) {
        self.labels = labels;
    }

    pub fn finish(mut self, outcome: McpOutcome, result_bytes: usize) {
        self.completed = true;
        self.telemetry.record(
            &self.request_id,
            self.labels,
            outcome,
            self.started.elapsed().as_secs_f64(),
            result_bytes,
        );
    }
}

impl Drop for McpObservation {
    fn drop(&mut self) {
        if !self.completed {
            self.telemetry.record(
                &self.request_id,
                self.labels,
                McpOutcome::Cancelled,
                self.started.elapsed().as_secs_f64(),
                0,
            );
        }
    }
}

pub fn labels_for(
    protocol: &'static str,
    method: Option<&str>,
    name: Option<&str>,
) -> McpMetricLabels {
    let method = safe_method(method);
    McpMetricLabels {
        protocol,
        operation: operation(method),
        method,
        name: safe_name(method, name),
    }
}

fn operation(method: &str) -> &'static str {
    match method {
        "server/discover" => "discovery",
        "initialize" => "initialization",
        "tools/list" => "listing",
        "tools/call" => "tool_call",
        "resources/list" | "resources/templates/list" | "resources/read" => "resource",
        "tasks/get" | "tasks/update" | "tasks/cancel" => "task",
        _ => "unknown",
    }
}

fn safe_method(method: Option<&str>) -> &'static str {
    match method {
        Some("initialize") => "initialize",
        Some("server/discover") => "server/discover",
        Some("tools/list") => "tools/list",
        Some("tools/call") => "tools/call",
        Some("resources/list") => "resources/list",
        Some("resources/templates/list") => "resources/templates/list",
        Some("resources/read") => "resources/read",
        Some("tasks/get") => "tasks/get",
        Some("tasks/update") => "tasks/update",
        Some("tasks/cancel") => "tasks/cancel",
        _ => "unknown",
    }
}

fn safe_name(method: &str, name: Option<&str>) -> &'static str {
    if method == "resources/read" {
        return match name {
            Some("ui://artifact-mcp/review") => "review_app",
            Some(uri) if uri.starts_with("artifact://") && uri.ends_with("/thumbnail") => {
                "artifact_thumbnail"
            }
            Some(uri) if uri.starts_with("artifact://") => "artifact_content",
            _ => "unknown",
        };
    }
    if method.starts_with("tasks/") {
        return "preview_regeneration";
    }
    if method != "tools/call" {
        return "none";
    }
    match name {
        Some("publish_artifact") => "publish_artifact",
        Some("publish_bundle") => "publish_bundle",
        Some("list_artifacts") => "list_artifacts",
        Some("delete_artifact") => "delete_artifact",
        Some("update_artifact") => "update_artifact",
        Some("set_visibility") => "set_visibility",
        Some("list_categories") => "list_categories",
        Some("set_category") => "set_category",
        Some("create_category") => "create_category",
        Some("delete_category") => "delete_category",
        Some("list_revisions") => "list_revisions",
        Some("create_share") => "create_share",
        Some("list_shares") => "list_shares",
        Some("revoke_share") => "revoke_share",
        Some("artifact_stats") => "artifact_stats",
        Some("submit_feedback") => "submit_feedback",
        Some("list_feedback") => "list_feedback",
        Some("resolve_feedback") => "resolve_feedback",
        Some("reopen_feedback") => "reopen_feedback",
        Some("read_artifact") => "read_artifact",
        Some("patch_artifact") => "patch_artifact",
        Some("regenerate_artifact_preview") => "regenerate_artifact_preview",
        _ => "unknown",
    }
}

fn result_size_bucket(bytes: usize) -> &'static str {
    match bytes {
        0..=1_024 => "le_1k",
        1_025..=16_384 => "le_16k",
        16_385..=262_144 => "le_256k",
        262_145..=1_048_576 => "le_1m",
        _ => "gt_1m",
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn metrics_keep_only_allowlisted_dimensions_and_bounded_sizes() {
        let telemetry = McpTelemetry::default();
        let mut observation = telemetry.begin();
        observation.set_labels(labels_for(
            "2026-07-28",
            Some("tools/call"),
            Some("secret-from-an-argument"),
        ));
        observation.finish(McpOutcome::Success, 2_048);

        let output = telemetry.render_prometheus();
        assert!(output.contains("method=\"tools/call\",name=\"unknown\",outcome=\"success\""));
        assert!(output.contains("size=\"le_16k\""));
        assert!(!output.contains("secret-from-an-argument"));
    }

    #[test]
    fn dropping_an_unfinished_observation_counts_cancellation() {
        let telemetry = McpTelemetry::default();
        {
            let mut observation = telemetry.begin();
            observation.set_labels(labels_for(
                "2026-07-28",
                Some("tools/call"),
                Some("list_artifacts"),
            ));
        }
        assert!(
            telemetry
                .render_prometheus()
                .contains("name=\"list_artifacts\",outcome=\"cancelled\"")
        );
    }
}
